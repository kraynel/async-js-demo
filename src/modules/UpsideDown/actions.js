export function toggleCall() {
  return {
    type: 'TOGGLE_CALL'
  };
}

export function updateStatus(newStatus) {
  return {
    type: 'UPDATE_STATUS',
    payload: {
      status: newStatus
    }
  };
}
