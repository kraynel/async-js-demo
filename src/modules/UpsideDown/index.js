// @flow
export { default as upsideDownReducer } from './reducer';
export { default as saga } from './sagas';
export * from './actions';
export * from './selectors';
