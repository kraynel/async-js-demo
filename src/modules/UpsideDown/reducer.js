// @flow

export type UpsideDownStateType = {
  status: string
};

const INITIAL_STATE = {
  status: 'In real world'
};

/**
* @param {Object} state - Default application state
* @param {Object} action - Action from action creator
* @returns {Object} New state
*/
export default (
  state: UpsideDownStateType = INITIAL_STATE,
  action
): UpsideDownStateType => {
  switch (action.type) {
    case 'UPDATE_STATUS':
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
