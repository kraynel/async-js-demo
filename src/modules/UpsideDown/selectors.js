// @flow

export const statusSelector = (state): string => state.upsideDown.status;
