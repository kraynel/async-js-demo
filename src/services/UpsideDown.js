class UpsideDown {
  canBeKilled = false;

  enter = callback => {
    console.info('Entering UpsideDown');
    setTimeout(callback, 1000);
  };

  findDemogorgon = callback => {
    console.info('Looking for demogorgon');
    setTimeout(() => {
      this.canBeKilled = true;
      callback('JUNKYARD');
    }, 1000);
  };

  killDemogorgon = callback => {
    console.info('Trying to kill the demogorgon');
    setTimeout(() => {
      if (this.canBeKilled) {
        console.log('Demogorgon is dead!');
        callback('SUCCESS');
        return;
      }
      console.log('You are dead!');
      callback('DEAD');
    }, 1000);
  };
}

export default UpsideDown;
