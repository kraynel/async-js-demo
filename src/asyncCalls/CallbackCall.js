import React, { Component } from 'react';
import UpsideDown from '../services/UpsideDown';

export default class CallbackCall extends Component {
  constructor(props) {
    super(props);
    this.upsideDown = new UpsideDown();
    this.state = {
      strangerStatus: 'In real world'
    };
  }

  toggleCall = () => {
    this.upsideDown.enter(() => {
      this.setState({ strangerStatus: 'In upside down' });

      this.upsideDown.findDemogorgon(location => {
        this.setState({ strangerStatus: `Demogorgon in ${location}` });

        this.upsideDown.killDemogorgon(result => {
          if (result === 'SUCCESS') {
            this.setState({ strangerStatus: 'Demogorgon dead ️☠️' });
            return;
          }

          this.setState({ strangerStatus: 'You are dead ☠️' });
        });
      });
    });
  };

  render() {
    return (
      <div>
        <button onClick={this.toggleCall}>Callback Call</button>
        <span>{this.state.strangerStatus}</span>
      </div>
    );
  }
}
