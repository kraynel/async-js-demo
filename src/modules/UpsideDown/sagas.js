// @flow

import { all, takeEvery, call, put } from 'redux-saga/effects';
import UpsideDownPromise from '../../services/UpsideDownPromise';

import { updateStatus } from './actions';

function* toggleCall(): Generator<*, *, *> {
  const upsideDown = new UpsideDownPromise();
  yield call([upsideDown, upsideDown.enter]);
  yield put(updateStatus('In under world'));

  const place = yield call([upsideDown, upsideDown.findDemogorgon]);
  yield put(updateStatus(`Found demogorgin in ${place}`));

  const result = yield call([upsideDown, upsideDown.killDemogorgon]);
  if (result === 'SUCCESS') {
    yield put(updateStatus(`Demogorgon is dead!`));
  } else {
    yield put(updateStatus(`You are dead`));
  }
}

export default function*(): Generator<*, *, *> {
  yield all([takeEvery('TOGGLE_CALL', toggleCall)]);
}
