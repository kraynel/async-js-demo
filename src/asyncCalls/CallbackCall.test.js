import React from 'react';
import renderer from 'react-test-renderer';

import UpsideDown from '../services/UpsideDown';
import CallbackCall from './CallbackCall';

jest.mock('../services/UpsideDown');

describe('CallbackCall', () => {
  it('should start in the real world', () => {
    const component = renderer.create(<CallbackCall />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should go in the upside down', () => {
    UpsideDown.mockImplementation(() => {
      return {
        enter: callback => callback(),
        findDemogorgon: () => {},
        killDemogorgon: () => {}
      };
    });

    const component = renderer.create(<CallbackCall />);
    component.getInstance().toggleCall();
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should find the demogorgon', () => {
    UpsideDown.mockImplementation(() => {
      return {
        enter: callback => callback(),
        findDemogorgon: callback => callback('SOME PLACE'),
        killDemogorgon: () => {}
      };
    });

    const component = renderer.create(<CallbackCall />);
    component.getInstance().toggleCall();
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should kill the demogorgon', () => {
    UpsideDown.mockImplementation(() => {
      return {
        enter: callback => callback(),
        findDemogorgon: callback => callback('SOME PLACE'),
        killDemogorgon: callback => callback('SUCCESS')
      };
    });

    const component = renderer.create(<CallbackCall />);
    component.getInstance().toggleCall();
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fail to kill the demogorgon', () => {
    UpsideDown.mockImplementation(() => {
      return {
        enter: callback => callback(),
        findDemogorgon: callback => callback('SOME PLACE'),
        killDemogorgon: callback => callback('FAILURE')
      };
    });

    const component = renderer.create(<CallbackCall />);
    component.getInstance().toggleCall();
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
