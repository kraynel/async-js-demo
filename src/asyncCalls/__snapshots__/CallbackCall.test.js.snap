// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallbackCall should fail to kill the demogorgon 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Callback Call
  </button>
  <span>
    You are dead ☠️
  </span>
</div>
`;

exports[`CallbackCall should find the demogorgon 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Callback Call
  </button>
  <span>
    Demogorgon in SOME PLACE
  </span>
</div>
`;

exports[`CallbackCall should go in the upside down 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Callback Call
  </button>
  <span>
    In upside down
  </span>
</div>
`;

exports[`CallbackCall should kill the demogorgon 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Callback Call
  </button>
  <span>
    Demogorgon dead ️☠️
  </span>
</div>
`;

exports[`CallbackCall should start in the real world 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Callback Call
  </button>
  <span>
    In real world
  </span>
</div>
`;
